### PR TITLE
GBA Save: Add support for 128kB SRAM and FlashROM saving (bootlegs)

### PR DESCRIPTION
**Pull request by metroid-maniac**
Add support for 128kB SRAM saving. Used by some Pokemon and Super Mario Advance 4 bootlegs. GBATA-style SRAM bank switching is also supported for support with SRAM-patched ROMs for older flashcarts.
Add support for rewritable ROM using flash commands. Used by "batteryless bootlegs" for saving.  ROM overlay is stored in the .sav file instead of SRAM, as these carts do not have batteries to retain SRAM contents persistently. SRAM is still used temporarily during gameplay.
Add some toggles to the Qt frontend to configure these parameters. 
Patches for batteryless ROMs can be found at https://github.com/acocalypso/batteryless-patches , Super Mario Advance 4 is of  interest as it also uses 128KB SRAM and will not save properly unless this is correctly emulated.

Relates to issue #2517

### CODE REVIEW

File: src/gba/overrides.c [hunk 1]
```diff
@@ -247,6 +247,9 @@ bool GBAOverrideFind(const struct Configuration* config, struct GBACartridgeOver
 			} else if (strcasecmp(savetype, "SRAM512") == 0) {
 				found = true;
 				override->savetype = SAVEDATA_SRAM512;
+			} else if (strcasecmp(savetype, "SRAM1M)") == 0) {
```
Comment (endrift) [inline src/gba/overrides.c:250]:
```suggestion
			} else if (strcasecmp(savetype, "SRAM1M") == 0) {
```

File: include/mgba/internal/gba/flashrom.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+#include <mgba-util/common.h>
+#include <mgba-util/memory.h>
+
+CXX_GUARD_START
+
+enum FlashROMType {
+	FLASHROM_NONE,
+	FLASHROM_22XX,
+	FLASHROM_INTEL
+};
```
Comment (endrift) [inline include/mgba/internal/gba/flashrom.h:13]: This should probably be in the Savedata type enum instead. Actually, most of or all of this should be rolled into the savedata struct.

File: include/mgba/internal/gba/flashrom.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	FLASHROM_INTEL_ERASE,
+	FLASHROM_INTEL_PROGRAM,
+	FLASHROM_INTEL_STATUS
+};
+
+enum {
+	FLASHROM_BLOCK_SIZE = 0x10000,
+};
+
+typedef uint8_t FlashROMBlock[FLASHROM_BLOCK_SIZE];
```
Comment (endrift) [inline include/mgba/internal/gba/flashrom.h:43]: Typedefs for a sized array are probably a bad idea. Certainly confusing.

File: src/platform/qt/OverrideView.ui [hunk 4]
```diff
@@ -204,6 +209,39 @@
...
+          </item>
+          <item row="3" column="0">
+           <widget class="QLabel">
+            <property name="text">
+             <string>FlashROM (bootlegs)</string>
+            </property>
+           </widget>
+          </item>
+          <item row="3" column="1">
+           <widget class="QComboBox" name="flashrom">
```
Comment (endrift) [inline src/platform/qt/OverrideView.ui:227]: This should just be more fields in the save types.

**Review by endrift — CHANGES_REQUESTED** [review]
I haven't managed to review much of the actual implementation yet, but to start with you should roll a lot of the novel structs and enums and fields into the existing savedata struct. The flash type does not need to be a separate enum, and a lot of stuff gets simplified if it's rolled into the existing structures.

Also, it crashed when I tried to use it. I'll look into why.

File: include/mgba/internal/gba/flashrom.h (hunk 2 above)
Comment (metroid-maniac) [reply include/mgba/internal/gba/flashrom.h:13]: I had a reason for separating it out. SMA4 needs 1M SRAM and will give warnings if 512K SRAM is present, but Pokemon Mystery Dungeon will corrupt its save if 1M SRAM is present and 512K SRAM is needed. Separating it out seemed reasonable.

In practice, I think there will only be 4 needed FlashROM configs for maximal compatibility (22xx 512K, 22xx 1M, 1M, Intel 512K, Intel 1M) so maybe it's better to just enumerate them in the normal save list.

**Comment by metroid-maniac** [comment]
Going back to the drawing board with this one

**Pull request closed without merging**